sysl2/sysl/importer/grammar.go

```diff
@@ -0,0 +1,205 @@
+			}
+		}
+	}
+	var baseType Type = &choiceType
+	return baseType
+}
+
+func (g *Grammar) quantifyUnionType(fieldName, fieldType string, quant *ebnfGrammar.Quantifier) Field {
+	if quant == nil {
+		return Field{
+			Name: fieldName,
+			Type: &StandardType{
+				name: fieldType,
+			}}
+	}
+	switch quant.Union.(type) {
+	case *ebnfGrammar.Quantifier_Optional:
+		fieldType = fmt.Sprintf("%s?", fieldType)
+	case *ebnfGrammar.Quantifier_ZeroPlus:
+		fieldType = fmt.Sprintf("sequence of %s?", fieldType)
+	case *ebnfGrammar.Quantifier_OnePlus:
+		fieldName = fmt.Sprintf("%s(1..)", fieldName)
+	default:
+		g.logger.Warnf("unexpected quantifier type")
+	}
+	return Field{
+		Name: fieldName,
+		Type: &StandardType{
+			name: fieldType,
+		}}
+}
+
+func (g *Grammar) createRule(rule *ebnfGrammar.Rule, name string) Type {
+	ruleType := StandardType{
+		name: name,
+	}
+	if !g.visitedMap[name] {
+		g.visitedMap[name] = true
+	}
+	for _, seq := range rule.Choices.Sequence {
+		for _, term := range seq.Term {
+			switch x := term.Atom.Union.(type) {
+			case *ebnfGrammar.Atom_String_:
+				// hit the leaf node
+				// return
+			case *ebnfGrammar.Atom_Choices:
+				choiceField := g.quantifyUnionType("__Choice_"+name, "__Choice_"+name, term.Quantifier)
+				ruleType.Properties = append(ruleType.Properties, choiceField)
+
+				// visit the unvisited node
+				if !g.visitedMap["__Choice_"+name] {
+					childChoiceType := g.createChoice(x.Choices, "__Choice_"+name)
+					g.types.types = append([]Type{childChoiceType}, g.types.types...)
+				}
+			case *ebnfGrammar.Atom_Rulename:
+				var childRuleType Type
+				typeString := "string"
+				if _, exists := g.grammar.Rules[x.Rulename.GetName()]; exists {
+					typeString = x.Rulename.Name
+				}
+				ruleField := g.quantifyUnionType(x.Rulename.Name, typeString, term.Quantifier)
+				ruleType.Properties = append(ruleType.Properties, ruleField)
+
+				// visit the unvisited node
+				if !g.visitedMap[x.Rulename.GetName()] {
+					if rule, exists := g.grammar.Rules[x.Rulename.GetName()]; exists {
+						childRuleType = g.createRule(g.grammar.Rules[rule.Name.Name], rule.Name.Name)
+						g.types.types = append([]Type{childRuleType}, g.types.types...)
+					}
+				}
+			default:
+				g.logger.Warnf("unexpected term type")
+			}
+		}
+	}
+	var baseType Type = &ruleType
+	return baseType
+}
+
+func (g *Grammar) writeSysl() (string, error) {
+	g.grammar = parser.ParseEBNF(g.grammarParam.grammarInput, g.grammarParam.grammarName,
+		g.grammarParam.startRule)
+	rule := g.grammar.Rules[g.grammarParam.startRule]
+	if rule == nil {
+		return "", fmt.Errorf("start rule not found")
+	}
+	var outputData OutputData
+	var syslBytes bytes.Buffer
+	syslWriter := newWriter(&syslBytes, g.logger)
+	switch g.grammarParam.startRule {
+	case `goFile`:
+		outputData = OutputData{
+			AppName: "Go",
+		}
+	case `javaFile`:
+		outputData = OutputData{
+			AppName: "Java",
+		}
+	default:
+		return "", fmt.Errorf("unexpected entry rule")
```

Comment:
The whole point of this feature is so new grammars don't need sysl files, why are you limiting support to go and java?!!?!?

Reply:
ok